exchanges/kucoin/kucoin_websocket.go

```diff
@@ -502,7 +465,30 @@ func (ku *Kucoin) processFuturesMarkPriceAndIndexPrice(respData []byte, instrume
 	return nil
 }
 
-func (ku *Kucoin) processFuturesOrderbookLevel5(respData []byte, instrument string) error {
+// ensureFuturesOrderbookSnapshotLoaded makes sure an initial futures orderbook snapshot is loaded
+func (ku *Kucoin) ensureFuturesOrderbookSnapshotLoaded(symbol string) error {
+	fetchedFuturesOrderbookMutex.Lock()
+	defer fetchedFuturesOrderbookMutex.Unlock()
+	if fetchedFuturesOrderbook[symbol] {
+		return nil
+	}
+	fetchedFuturesOrderbook[symbol] = true
```

Comment:
Not huge, but shouldn't this be at the end in the event of an error in the 4 functions below? Its only ensuring that things run once per pair otherwise

Reply:
Hmmm.
So my issue with this is the back-off we'd need.
If we get orderbook updates, and we haven't loaded the snapshot, and we don't say we've loaded it before we error, we're going to keep looping around in an expensive stream-rate-driven loop.

If it's okay, I'll swerve fixing this issue until I come back and deal with this better entirely.

I think you already noted: I didn't introduce this issue you're citing; it's just been moved from line 294 to here.

Reply:
Yeah it can be added to the pile. Its less about subscription templating